core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala

```diff
@@ -28,12 +31,18 @@ private[spark] class JobWaiter[T](
     resultHandler: (Int, T) => Unit)
   extends JobListener {
 
+  private val promise = Promise[Unit]
```

Comment:
Stepping back. If we are using promise any ways, why do we need a separate variable called "jobFinished"? The promise is sufficient in keeping the state of whether the job has finished or not.